io/io/src/TFileMergeServer.cxx

```diff
@@ -0,0 +1,348 @@
+   }
+
+   Bool_t NeedMerge(Float_t clientThreshold)
+   {
+      if (fClients.size() == 0) {
+         return false;
+      }
+
+      // Calculate average and rms of the time between the last 2 contacts.
+      Double_t sum = 0;
+      Double_t sum2 = 0;
+      for (unsigned int c = 0; c < fClients.size(); ++c) {
+         sum += fClients[c].fTimeSincePrevContact;
+         sum2 += fClients[c].fTimeSincePrevContact * fClients[c].fTimeSincePrevContact;
+      }
+      Double_t avg = sum / fClients.size();
+      Double_t sigma = sum2 ? TMath::Sqrt(sum2 / fClients.size() - avg * avg) : 0;
+      Double_t target = avg + 2 * sigma;
+      TTimeStamp now;
+      if ((now.AsDouble() - fLastMerge.AsDouble()) > target) {
+         return true;
+      }
+      Float_t cut = clientThreshold * fClients.size();
+      return fClientsContact.CountBits() > cut || fNClientsContact > 2 * cut;
+   }
+
+   Bool_t Merge()
+   {
+      // Merge the current inputs into the output file.
+
+      // Remove object that can *not* be incrementally merged
+      // and will *not* be reset by the client code.
+      R__DeleteObject(fMerger.GetOutputFile(), false);
+
+      for (unsigned int f = 0; f < fClients.size(); ++f) {
+         fMerger.AddFile(fClients[f].fFile);
+      }
+
+      Bool_t result = fMerger.PartialMerge(TFileMerger::kAllIncremental);
+
+      // Remove any 'resetable' object (like TTree) from the input file
+      // so that they will not be re-merged. Keep only the object that
+      // always need to be re-merged (Histograms).
+
+      for (unsigned int f = 0; f < fClients.size(); ++f) {
+         if (fClients[f].fFile) {
+            R__DeleteObject(fClients[f].fFile, true);
+         } else {
+            // We back up the file (probably due to memory constraint)
+            TFile *file = TFile::Open(fClients[f].fLocalName, "UPDATE");
+            // Remove object that can be incrementally merged and
+            // will be reset by the client code.
+            R__DeleteObject(file, true);
+            file->Write();
+            delete file;
+         }
+      }
+      fLastMerge = TTimeStamp();
+      fNClientsContact = 0;
+      fClientsContact.Clear();
+
+      return result;
+   }
+
+   void RegisterClient(UInt_t clientId, TFile *file)
+   {
+      ++fNClientsContact;
+
+      fClientsContact.SetBitNumber(clientId);
+
+      if (fClients.size() < clientId + 1) {
+         fClients.push_back(ClientInfo(fFilename, clientId));
+      }
+      fClients[clientId].Set(file);
+   }
+};
+
+} // unnamed namespace
+
+namespace ROOT {
+namespace Experimental {
+
+std::shared_ptr<TFileMergeClient> TFileMergeServer::GetClient()
+{
+   std::shared_ptr<TFileMergeClient> client(new TFileMergeClient(*this));
+   fClients.push_back(client);
+   return client;
+}
+
+void TFileMergeServer::Listen()
+{
+   std::unique_lock<std::mutex> wlock(fWriteMutex);
+
+   THashTable mergers;
+   bool cache = false;
+
+   static Int_t clientId = 0;
```

Comment:
Yes, as Philippe mentioned before, this is an artifact of the imported code. Not something for fixing immediately. Later I will remove the ThreadFileMerger and ClientInfo structs and include the functionality into TFileMergeServer.